src/libexpr/primops.cc

```diff
@@ -1009,20 +1009,16 @@ static void prim_toFile(EvalState & state, const Pos & pos, Value * * args, Valu
 }
 
 
-static void prim_filterSource(EvalState & state, const Pos & pos, Value * * args, Value & v)
-{
-    PathSet context;
-    Path path = state.coerceToPath(pos, *args[1], context);
-    if (!context.empty())
-        throw EvalError(format("string '%1%' cannot refer to other paths, at %2%") % path % pos);
-
-    state.forceValue(*args[0]);
-    if (args[0]->type != tLambda)
-        throw TypeError(format("first argument in call to 'filterSource' is not a function but %1%, at %2%") % showType(*args[0]) % pos);
-
-    path = state.checkSourcePath(path);
-
-    PathFilter filter = [&](const Path & path) {
+static void addPath(EvalState & state, const Pos & pos, const string & name, const Path & path_,
+    Value * filterFun, bool recursive, const Hash & expectedHash, Value & v)
+{
+    /* !!! Possibly unsafe to bypass restricted eval even with
+       expected hash... you can confirm some file with known hash
+       exists at an arbitrary path. */
+    const auto path = expectedHash ?
+        path_ :
+        state.checkSourcePath(path_);
```

Comment:
This looks wrong. In `--restricted-eval`, the path should be checked even if a hash is given. In `--pure-eval`, the path doesn't need to be checked but it requires a hash.